Add support for hierarchy/measure display folder

jbarefoot opened the pull request:
For reference from the olap4j spec, here are the two properties I'm trying to surface in the object model:
http://www.olap4j.org/olap4j_fs.html
HIERARCHY_DISPLAY_FOLDER
MEASURE_DISPLAY_FOLDER

...and here's the commit message for how it was done:
- add interface to support this without breaking other implementations e.g. Mondrian's
- add property to measure and hierarchy object models from the standard XMLA properties for folders

The key here is to not break Mondrian's implementations of Measure and Hierarchy.  Note also that apparently Mondrian has the notion of a "measureGroup"--I'm not sure how that differs from MEASURE_DISPLAY_FOLDER.  


Motivation and Saiku:

The goal of adding display folders is twofold:  1.  Our (AtScale) customers want to use olap4j for custom app development, and want folders.  2.  One of our customers is migrating from SSAS to AtScale, and wants to use a modified version of Saiku against both, and of course want folders to work.  I have also added display folder to Saiku's model as you can see in this fork diff:
https://github.com/OSBI/saiku/compare/development...AtScaleInc:atscale_add_folder_support

Usage is in the ObjectUtil convert* methods:
https://github.com/OSBI/saiku/compare/development...AtScaleInc:atscale_add_folder_support#diff-477feb83d9e3154910a1b94742b70195R115

Until DisplayFolder is available in an updated olap4j jar in some repo, the saiku fork has the built olap4j artifacts checked in temporarily so that one can do a local mvn install to test.  The goal is to also get the javascript modified for the front-end of saiku so that hierarchy folders work in its UI. 

(Sorry for the long PR description, but I thought folks might be interested in knowing there's actually a plan here.)

julianhyde commented:
I am concerned that once the interface is added, we would not be able to remove it. Rather than adding an interface, would it be sufficient to just add a property? You could get the property from the member or the hierarchy via the {{getPropertyValue(Property)}} method.

In olap4j version 2, we are at liberty to add new methods to `Measure` and `Hierarchy`, so we could add a `getDisplayFolder()` method. In fact, in the olap4j2 branch we have already added that method to `Hierarchy`. There is a question of how soon clients and servers would upgrade to olap4j2. I am hoping that clients such as Saiku would be able to talk to both 1.x and 2.x servers, or perhaps use a bridging provider that makes everything look like olap4j 2.0.

By returning the "HIERARCHY_DISPLAY_FOLDER" and "MEASURE_DISPLAY_FOLDER" properties, a server supporting olap4j 1.3 or 2.0 would be able to get the information to any client.

julianhyde commented:
By the way, measure groups are collections of measures that have the same fact table. They have nothing to do with folders, or how measures are displayed. The end user is not aware of measure groups, except for the fact that measures have different granularity.

jbarefoot commented:
Thanks for the feedback and info on measure groups.  Regarding properties:  I looked at this when I made my changes.  I would be happy to use `getPropertyValue(Property)` but I don't see any way to do it without still adding a new interface.  Mostly this is because `getPropertyValue(Property)` is on the `Member` interface but `Hierarchy` doesn't extend that.  Also, all of the implementation classes have package scope and thus aren't visible outside the project, so I couldn't just add it to `XmlaOlap4jHierarchy` directly.

Complicating matters, I can't e.g. just move the property methods up from `Member` into the `MetadataElement` interface (where I feel they really belong), because `Hierarchy` *does* extend that, and Mondrian in turn implements `Hierarchy`, so that would break Mondrian.  The only alternative thing I can think of to do would be to add a `PropertyMetadata` interface with just the property methods, refactor `Member` to extend that, and make `XmlaOlap4jHierarchy` implement `PropertyMetadata`.  I would still be doing an `intanceof` in Saiku's code to use  `PropertyMetadata`, however maybe `PropertyMetadata` is more acceptable as a supported interface than `DisplayFolder`.  What do you think?

julianhyde commented:
I understand now why you need to add the `DisplayFolder` interface. The real solution is to extend the `Hierarchy` interface, and we can do that in olap4j version 2. (Breaking changes are allowed in a major release.)

For the 1.x code line, I suggest that you add `DisplayFolder` and make it deprecated. It will remain in the 1.x code line and will not be present in 2.0. If, say, Saiku wants to work against both olap4j 1.x and olap4j 2.0 it will have to use `Class.forName("org.olap4j.metadata.DisplayFolder")` or similar.

Would this be acceptable?

julianhyde commented:
One difference between master and olap4j2 is that on the latter I moved files into standard maven structure: src/main/java rather than src/java, src/test/java rather than testsrc/java. Note that I added a few lines to build.xml so that ant (and subfloor) knows what is going on.

Do people think it is worth moving files around to accomplish that? IIRC, a few `git mv` commands are sufficient. And it should make it easier to merge between branches.

Other than that, looks good - nice work, @buggtb.

jbarefoot commented:
Sounds good to me.  I added a javadoc tag to `DisplayFolder` to make it deprecated with a comment, and added the missing license header.  I'll have a look at what to do for olap4j 2.0 and and Saiku.

julianhyde commented:
Fixed in master branch in commit 99ec7a4b23a0ab3c682098ee707eb9605f387368. Has yet to be merged to olap4j2 branch.

julianhyde commented:
Adjusted the fix in the olap4j2 branch in commit f36e83f993435f956c4cfd5883f58b0354f88b02.
